### ISSUE TITLE
Please document how to generate a commonJS/systemJS/AMD JS module from python using transcrypt for loading using require.js or node/etc.

### ISSUE THREAD

**Issue by shlomif** [QQuick/Transcrypt#495]
Hi!

Please document how to generate a commonJS/systemJS/AMD JS module from python using transcrypt for loading using require.js or node/etc.

I saw this discussion - https://github.com/QQuick/Transcrypt/issues/94 - but I do not understand how to do it based on that.

**Comment by JdeH** [comment]
I don't currently don't know enough from JS module systems to describe this properly.
Can anyone give it  a try, e.g. with some examples.
I am more a Python than a JS guy...

**Comment by doconix** [comment]
I know enough (I think :) to write it up.  However, there are a couple things that transcrypt needs before it can happen.  

1. We need to be able to do JS imports and exports at the top level of the file.  Right now, transcrypt doesn't provide a way to place code outside the containing function.  The JS (ES6) standard for the import statement says it has to happen *outside* any function--in the main space of the .js file.

2. We need to be able to disable placing the function in the global `window` object (very last line of the generated JS).  In other words, we need to disable calling the function entirely.  Just create the function and let the `exports` code (that goes in the top level of the file (#1 above) export the function according to the AMD, CommonJS, etc. ways.

From what I can see, the two above items are everything needed to make module creation happen.

**Comment by shlomif** [comment]
On Thu, 08 Mar 2018 10:10:31 -0800
Conan Albrecht <notifications@github.com> wrote:

> I know enough (I think :) to write it up.  However, there are a couple things
> that transcrypt needs before it can happen.  
> 
> 1. We need to be able to do JS imports and exports at the top level of the
> file.  Right now, transcrypt doesn't provide a way to place code outside the
> containing function.  The JS (ES6) standard for the import statement says it
> has to happen *outside* any function--in the main space of the .js file.
> 
> 2. We need to be able to disable placing the function in the global `window`
> object (very last line of the generated JS).  In other words, we need to
> disable calling the function entirely.  Just create the function and let the
> `exports` code (that goes in the top level of the file (#1 above) export the
> function according to the AMD, CommonJS, etc. ways.
> 
> From what I can see, the two above items are everything needed to make module
> creation happen.
> 

thanks for the insights, Conan!

-- 
-----------------------------------------------------------------
Shlomi Fish       http://www.shlomifish.org/
http://www.shlomifish.org/humour/bits/Google-Discontinues-Services/

Yesterday I asked one of my students if she knew what an encyclopedia is, and
she said: “Is it something like Wikipedia?”.
        — http://twitter.com/alisonclement/status/8421314259

Please reply to list if it's a mailing list post - http://shlom.in/reply .

**Comment by JdeH** [comment]
@doconix 

Could you manually modify a simple, generated JS file, e.g. hello.js (from hello.py) to lay down exactly what is deed.

You can use the `-p .user` switch to avoid the main function to be called.
If you manually insert the needed export / import code, we can later on create a facility to insert it at will, e.g. as a free to choose prologue and epilogue by means of a pragma.

**Comment by doconix** [comment]
Will do!

**Comment by doconix** [comment]
Wow.  So I totally forgot that I already did this with an earlier project and published the plugin on `npm`.  I'm getting old.

Here's the repo: https://github.com/doconix/autotimers
In particular, check out the exports at the bottom of the main file: https://github.com/doconix/autotimers/blob/master/src/main.py#L117

The code is a little different than the common pattern used in JS.  Most modules follow this pattern: https://github.com/umdjs/umd.  I modified it because using Python code syntax makes it quite a bit simpler and easier.  To my view, my code does the exact thing, just with a refactored way.

The good news for @JdeH is no changes are needed in Transcrypt.  Although I will note that using `npm` modules from Transcrypt code is difficult because of #1 above.  The workaround is to use the deprecated `require()` function.  Since the npm repository seems to be the future of JS modules, it might be a good thing to support `import` in Transcrypt.  The issue deserves its own page instead of this paragraph, but I'll leave any development at @JdeH discretion.

**Comment by JdeH** [comment]
@doconix 
What do you mean by #1? (It links to the very first issue, which is probably not what is meant.

Do you mean the following point of @shlomif :
```
1. We need to be able to do JS imports and exports at the top level of the
 file.  Right now, transcrypt doesn't provide a way to place code outside the
 containing function.  The JS (ES6) standard for the import statement says it
 has to happen *outside* any function--in the main space of the .js file.
```

So if Transcrypt would offer the possibility to have code before and after the main function definition, that would help, right?

**Comment by shlomif** [comment]
On Thu, 08 Mar 2018 15:17:33 -0800
Conan Albrecht <notifications@github.com> wrote:

> Wow.  So I totally forgot that I already did this with an earlier project and
> published the plugin on `npm`.  I'm getting old.
> 
> Here's the repo: https://github.com/doconix/autotimers
> In particular, check out the exports at the bottom of the main file:
> https://github.com/doconix/autotimers/blob/master/src/main.py#L117
> 

thanks, Conan! Will it be possible to write a tool to automate this process?



> The code is a little different than the common pattern used in JS.  Most
> modules follow this pattern: https://github.com/umdjs/umd.  I modified it
> because using Python code syntax makes it quite a bit simpler and easier.  To
> my view, my code does the exact thing, just with a refactored way.
> 
> The good news for @JdeH is no changes are needed in Transcrypt.  Although I
> will note that using `npm` modules from Transcrypt code is difficult because
> of #1 above.  The workaround is to use the deprecated `require()` function.
> Since the npm repository seems to be the future of JS modules, it might be a
> good thing to support `import` in Transcrypt.  The issue deserves its own
> page instead of this paragraph, but I'll leave any development at @JdeH
> discretion.
> 



-- 
-----------------------------------------------------------------
Shlomi Fish       http://www.shlomifish.org/
https://youtu.be/GoEn1YfYTBM - Tiffany Alvord - “Fall Together”

“The best time to plant a tree was 20 years ago. The second best time is now.”
    – A Reported Proverb

Please reply to list if it's a mailing list post - http://shlom.in/reply .

**Comment by JdeH** [comment]
Investigating adding ES6 modules to TS.
Testing with Chrome.
Still some trouble in paradise...

https://stackoverflow.com/questions/49190897/cant-get-export-as-default-to-work-on-chrome-version-64-0-3282-186-official-bu

**Comment by doconix** [comment]
Regarding "#1" above, we need to be able to add the `import` statement at the top of the following example.  It is outside the `dev` function below -- at the top level of the file.   

```
"use strict";
import $ as 'jquery';
// Transcrypt'ed from Python, 2017-11-17 21:55:04
function dev () {
    ...
}
dev ();
//# sourceMappingURL=extra/sourcemap/dev.js.map
```

**Comment by doconix** [comment]
I see from your SO question that you are working on exporting ES6 modules.  That's the right approach to work for.

The solution I referenced above (https://github.com/doconix/autotimers/blob/master/src/main.py#L117) implements the "universal module definition".  It supports the different methods that have been competing for the last few years: CommonJS, AMD (requirejs), etc.  ES6 modules are the result of this competition and is the future. 

So while my solution works well right now, the way you are going is the standards approach.  I should have suggested this rather than the universal one I did.  Just wanted to validate where you are headed with it.

**Comment by JdeH** [comment]
Your solution is very useful, since it addresses the currently popular modules systems.
However the "official" JS6 one is the one I want to include by default in Transcrypt.
I've been experimenting with it, and it alreay works, but it doesn't please me, since it is orthogonal to TS/Python imports.
What I'm after is a mode where every Python import automatically is a JS6 import.
So pervasive module loading on demand.
What I've been running in to is that the support in e.g. Chrome doesn't always seem to work right.
So I'm not sure it's all going to work out.
But trying...

**Comment by doconix** [comment]
That's beautiful.  Imports aren't fully into the browsers yet, so that's one possible reason it's difficult.  In any case, almost all people working with ES6+ code right now are transpiling back to 5 with webpack, babel, etc.  So in my mind, the real question is whether babel can handle the imports.  

I'm sure you know, but note also that the new `import()` function (not the statement) is being integrated into the JS standard right now.  So that's a possibility for making it work as well.

Tough to hit a moving target...

**Comment by metamarcdw** [comment]
I would definitely love to see the prologue/epilogue pragmas as a stopgap until the ES6 module stuff is done. Any pointers on how I could get started on that?

**Comment by JdeH** [comment]
I've already built in and removed it again, since it messed up the sourcemaps. (linenumbers are offset).
Currently the best thing to do is probably:

- If you want to avoid the main encapsulator function to be called and its result assigned to `window`, use the -p .user switch:

From the code:
`self.argParser.add_argument ('-p', '--parent', nargs='?', help = "object that will hold application, default is window. Use -p .none to generate orphan application, e.g. for use in node.js. Use -p .user to generate an application that has to be explicitly initialized by calling <application name> (), e.g. after the full page has loaded")`

- To add a prologue and epilogue, the best thing to do (since it's only temporary) is to read the generated .js file into a small postprocessing application and glue something before and after it, then write the result.

Releasing a pragma suggests it will be a permanent facility, which I want to avoid, because taking it out again will break user code.

**Comment by shlomif** [comment]
Can someone provide a script or whatever to automate the process as a gist or similar? I don't fully understand what to do.

**Comment by doconix** [comment]
@shlomif I scaled down my project and posted to https://github.com/doconix/transcrypt_module/.  Please see the src/main.py for the code on making a UMD.  

This should work with 1) browser, 2) AMD, and 3) CommonJS.  I have tested my original project with these three, but I've only had time today to test the first.  It should work with all three, though, because the code is essentially the same.

Note that this does NOT make the module work with the ES6 standard.  UMD and ES6 are not really compatible, from what I see.  An ES6 module would need to `export default ...`.  I'm not sure if that can be done with Transcrypt yet.

But in any case, the code I posted will work with today's tech: webpack, gulp, browserify, node, etc.

**Comment by doconix** [comment]
Also note that it needs to be built with specific transcrypt options.  I created the `build.py` file to do it the right way.

**Comment by shlomif** [comment]
On Tue, 13 Mar 2018 15:11:07 +0000 (UTC)
Conan Albrecht <notifications@github.com> wrote:

> @shlomif I scaled down my project and posted to
> https://github.com/doconix/transcrypt_module/.  Please see the src/main.py
> for the code on making a UMD.  
> 

many thanks - I shall take a look at it soon.

> This should work with 1) browser, 2) AMD, and 3) CommonJS.  I have tested my
> original project with these three, but I've only had time today to test the
> first.  It should work with all three, though, because the code is
> essentially the same.
> 
> Note that this does NOT make the module work with the ES6 standard.  UMD and
> ES6 are not really compatible, from what I see.  An ES6 module would need to
> `export default ...`.  I'm not sure if that can be done with Transcrypt yet.
> 
> But in any case, the code I posted will work with today's tech: webpack,
> gulp, browserify, node, etc.
> 



-- 
-----------------------------------------------------------------
Shlomi Fish       http://www.shlomifish.org/
UNIX Fortune Cookies - http://www.shlomifish.org/humour/fortunes/

Chuck Norris read the entire English Wikipedia in 24 hours. Twice.
    — http://www.shlomifish.org/humour/bits/facts/Chuck-Norris/

Please reply to list if it's a mailing list post - http://shlom.in/reply .

**Comment by shlomif** [comment]
@doconix : can you please add a LICENSE/etc. file to the repo so it won't be ARR? See https://github.com/shlomif/Freenode-programming-channel-FAQ/blob/master/FAQ.mdwn#i-want-to-release-my-code---which-open-source-licence-should-i-use . I hope I don't seem too needy and picky.

**Comment by doconix** [comment]
@shlomif Done

**Comment by shlomif** [comment]
On Tue, 13 Mar 2018 18:02:09 +0000 (UTC)
Conan Albrecht <notifications@github.com> wrote:

> @shlomif Done
> 
> 

awesome, thanks!

-- 
-----------------------------------------------------------------
Shlomi Fish       http://www.shlomifish.org/
Original Riddles - http://www.shlomifish.org/puzzles/

Barth's Distinction:  There are two types of people: those who divide people
into two types, and those who don't.
    — via fortune-mod.

Please reply to list if it's a mailing list post - http://shlom.in/reply .

**Comment by shlomif** [comment]
@doconix : it looks very good, thanks - I'll let you know what evolves next.

**Comment by shlomif** [comment]
@doconix : thanks again, the code is now used live at http://fc-solve.shlomifish.org/js-fc-solve/find-deal/ . There still is some polish to be done , but it is not related to transcrypt.

**Comment by JdeH** [comment]
ES6 modules are now available in the 3.7.1.alpha prerelease.

**Comment by shlomif** [comment]
On Mon, 09 Apr 2018 07:04:35 -0700
Jacques de Hooge <notifications@github.com> wrote:

> ES6 modules are now available in the 3.7.1.alpha prerelease.
> 

thanks!

-- 
-----------------------------------------------------------------
Shlomi Fish       http://www.shlomifish.org/
Emma Watson Factoids - http://shlom.in/emwatson-facts

Chuck Norris helps God help those that help themselves.
    — http://www.shlomifish.org/humour/bits/facts/Chuck-Norris/

Please reply to list if it's a mailing list post - http://shlom.in/reply .
